File: 1989-minimum-xor-sum-of-two-arrays/minimum-xor-sum-of-two-arrays.java
class Solution {
    public int minimumXORSum(int[] nums1, int[] nums2) {
        // dp with bitmasking , write recursive
        int states = 1 << nums1.length;
        int dp[] = new int[states];
        Arrays.fill(dp, -1);
        xorf(nums1, nums2, dp, 0, 0);
        System.out.println(Arrays.toString(dp));
        return dp[0];
    }

    public static int xorf(int nums1[], int nums2[], int dp[], int bitmask, int idx) {
        if (idx == nums1.length)
            return 0;
        if (dp[bitmask] != -1)
            return dp[bitmask];

        int ans = (int) 1e9;

        for (int i = 0; i < nums2.length; i++) {
            if ((bitmask & (1 << i)) > 0)
                continue;
            ans = Math.min((nums1[idx] ^ nums2[i]) + xorf(nums1, nums2, dp, (bitmask | (1 << i)), idx + 1),ans);
        }
        return dp[bitmask] = ans;
    }
}
//152 167 175 167 167 207 167 120 164 206 164 117 204 157 123 92 125 165 125 76 165 116 76 51 162 119 73 54 113 110 48 2147483647 

File: 1989-minimum-xor-sum-of-two-arrays/minimum-xor-sum-of-two-arrays.cpp
#include <bits/stdc++.h>
using namespace std;

#define ff first
#define ss second
#define pb push_back
#define MOD 1000000007
#define inf 1e18

int solve(int i, int mask, vector<vector<int>>& dp, vector<int>& nums1,
          vector<int>& nums2) {
    if (i == nums1.size()) {
        return 0;
    }
    if (dp[i][mask] != -1) {
        return dp[i][mask];
    }
    int res = INT_MAX;
    for (int k = 0; k < nums1.size(); k++) {
        if ((mask & (1 << k)) == 0) { // jth element is not used till now
            res = min(res, ((nums1[i] ^ nums2[k]) + solve(i + 1, mask | (1 << k),
                                                        dp, nums1, nums2)));
        }
    }
    return dp[i][mask] = res;
}

class Solution {
public:
    int minimumXORSum(vector<int>& nums1, vector<int>& nums2) {
        // since the given question hai n = 14 so i can apply the approach of dp
        // with bitmasking yrr

        int n = nums1.size();
        vector<vector<int>> dp(n + 1, vector<int>(1 << (n + 1), -1));
        return solve(0, 0, dp, nums1, nums2);
    }
};

#ifdef LOCAL
int32_t main() {
    freopen("output.txt", "w", stdout);
    cout << "Compiled!" << endl;
}
#endif
